components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java

```diff
@@ -445,26 +443,26 @@ private void removeExpiredSessionData(String sqlQuery) {
             boolean deleteCompleted = false;
             int totalDeletedEntries = 0;
             while (!deleteCompleted) {
-                statement = connection.prepareStatement(sqlQuery);
-                statement.setLong(1, currentTime);
-
-                int noOfDeletedRecords = statement.executeUpdate();
-                deleteCompleted = noOfDeletedRecords < deleteChunkSize;
-                totalDeletedEntries += noOfDeletedRecords;
-                if (log.isDebugEnabled()) {
-                    log.debug(String.format("Removed %d expired session records.",
-                            noOfDeletedRecords));
+                try (PreparedStatement statement = connection.prepareStatement(sqlQuery)){
+                    statement.setLong(1, currentTime);
+                    int noOfDeletedRecords = statement.executeUpdate();
+                    deleteCompleted = noOfDeletedRecords < deleteChunkSize;
+                    totalDeletedEntries += noOfDeletedRecords;
+                    // Commit the chunk deletion.
+                    IdentityDatabaseUtil.commitTransaction(connection);
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("Removed %d expired session records.", noOfDeletedRecords));
+                    }
                 }
-                IdentityDatabaseUtil.rollbackTransaction(connection);
             }
             if (log.isDebugEnabled()) {
-                log.debug(String.format("Deleted total of %d entries ", totalDeletedEntries));
+                log.debug(String.format("Deleted total of %d entries", totalDeletedEntries));
             }
         } catch (SQLException e) {
             IdentityDatabaseUtil.rollbackTransaction(connection);
-            log.error("Error while removing session data from the database for nano time " + currentTime, e);
+            log.error("Error while removing session data from the database for nano time: " + currentTime, e);
         } finally {
-            IdentityDatabaseUtil.closeAllConnections(connection, null, statement);
+            IdentityDatabaseUtil.closeAllConnections(connection, null, null);
```

Comment:
Let's remove the finally block and enclose the connection in a try with resource block.

Reply:
Fixed in a5f2647666a6f6cd722efacc141af549d9a14496.